source/custom/nidaqmx_service.custom.cpp

```diff
@@ -1,48 +1,97 @@
 #include <nidaqmx/nidaqmx_service.h>
+#include <server/async_operation.h>
+#include <server/async_processing_loop.h>
 #include <server/callback_router.h>
 
+#include <memory>
 #include <thread>
 
 namespace nidaqmx_grpc {
 
 using DoneEventCallbackRouter = nidevice_grpc::CallbackRouter<int32, TaskHandle, int32>;
 
+class DoneEventCallbackOperation : public nidevice_grpc::CompletionQueueElement {
```

Comment:
This "self-owning" pattern is pretty common in demos of async grpc:

1. `new` up the class but don't keep the pointer
2. Register with the CompletionQueue using `this` as the tag.
3. The CompletionQueue can trivially get back to the element by casting it and calling `process`.
4. In `process` the element either calls `delete this` or goes back to step 2.

Most implementations use a state machine and repeat steps 2-4. I'm always deleting and potentially allocating a new queue element. The big downside there is that I'm hitting the heap more, although that could be optimized in a couple different ways (using a pool; allocating the related steps as a block).

The "rationale" for my version:
1. Makes it easier to separate the common `FinishOperation` logic in totally separate classes.
2. The "state machine" works best when you continually reuse the same tag. But `AsyncNotifyWhenDone` always needs to be a separate tag. So we'd need multiple elements even if we consolidated.
3. Though the `shared_ptr` ownership is a little overkill for the strictly-async-grpc aspect of this implementation. It's really helpful for allowing the `callback` to participate in ownership. Since that can be called literally at any time, we need an ownership model that supports that.